pkg/solana/client/multinode_client.go

```diff
@@ -0,0 +1,306 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+
+	mn "github.com/smartcontractkit/chainlink-solana/pkg/solana/client/multinode"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/config"
+)
+
+type Head struct {
+	BlockHeight *uint64
+	BlockHash   *solana.Hash
+}
+
+func (h *Head) BlockNumber() int64 {
+	if !h.IsValid() {
+		return 0
+	}
+	// nolint:gosec
+	// G115: integer overflow conversion uint64 -&gt; int64
+	return int64(*h.BlockHeight)
+}
+
+func (h *Head) BlockDifficulty() *big.Int {
+	// Not relevant for Solana
+	return nil
+}
+
+func (h *Head) IsValid() bool {
+	return h.BlockHeight != nil && h.BlockHash != nil
+}
+
+var _ mn.RPCClient[mn.StringID, *Head] = (*MultiNodeClient)(nil)
+var _ mn.SendTxRPCClient[*solana.Transaction] = (*MultiNodeClient)(nil)
+
+type MultiNodeClient struct {
+	Client
+	cfg         *config.TOMLConfig
+	stateMu     sync.RWMutex // protects state* fields
+	subsSliceMu sync.RWMutex
+	subs        map[mn.Subscription]struct{}
+
+	// chStopInFlight can be closed to immediately cancel all in-flight requests on
+	// this RpcClient. Closing and replacing should be serialized through
+	// stateMu since it can happen on state transitions as well as RpcClient Close.
+	chStopInFlight chan struct{}
+
+	chainInfoLock sync.RWMutex
+	// intercepted values seen by callers of the rpcClient excluding health check calls. Need to ensure MultiNode provides repeatable read guarantee
+	highestUserObservations mn.ChainInfo
+	// most recent chain info observed during current lifecycle (reseted on DisconnectAll)
+	latestChainInfo mn.ChainInfo
+}
+
+func NewMultiNodeClient(endpoint string, cfg *config.TOMLConfig, requestTimeout time.Duration, log logger.Logger) (*MultiNodeClient, error) {
+	client, err := NewClient(endpoint, cfg, requestTimeout, log)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MultiNodeClient{
+		Client:         *client,
+		cfg:            cfg,
+		subs:           make(map[mn.Subscription]struct{}),
+		chStopInFlight: make(chan struct{}),
+	}, nil
+}
+
+// registerSub adds the sub to the rpcClient list
+func (m *MultiNodeClient) registerSub(sub mn.Subscription, stopInFLightCh chan struct{}) error {
+	m.subsSliceMu.Lock()
+	defer m.subsSliceMu.Unlock()
+	// ensure that the `sub` belongs to current life cycle of the `rpcClient` and it should not be killed due to
+	// previous `DisconnectAll` call.
+	select {
+	case <-stopInFLightCh:
+		sub.Unsubscribe()
+		return fmt.Errorf("failed to register subscription - all in-flight requests were canceled")
+	default:
+	}
+	// TODO: BCI-3358 - delete sub when caller unsubscribes.
+	m.subs[sub] = struct{}{}
+	return nil
+}
+
+func (m *MultiNodeClient) Dial(ctx context.Context) error {
```

Comment:
You should put a comment here as it's weird to have a function that just returns nil. Is this a TODO to implement later? Is this functionality not needed for Solana?

Reply:
It's not needed for Solana, I added a comment :)